### ISSUE TITLE
Node Transformer throws class cast exception

### ISSUE THREAD

**Issue by nadeeshaan** [ballerina-platform/ballerina-lang#25125]
**Description:**
Consider the following source snippet,

```
public function testFunction() {
    int[] numberList = [1, 2, 3];

    foreach i
}

function doSomeTask() {
    // todo: here goes a certain task
}

```

For the above, Node transformer throws a CCE as follows,

```
Caused by: java.lang.ClassCastException: io.ballerinalang.compiler.syntax.tree.FunctionalBindingPatternNode cannot be cast to io.ballerinalang.compiler.syntax.tree.CaptureBindingPatternNode
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.getBLangVariableNode(BLangNodeTransformer.java:3640)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createBLangVarDef(BLangNodeTransformer.java:2434)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:1338)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:418)
	at io.ballerinalang.compiler.syntax.tree.ForEachStatementNode.apply(ForEachStatementNode.java:62)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.generateAndAddBLangStatements(BLangNodeTransformer.java:3670)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:1325)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:418)
	at io.ballerinalang.compiler.syntax.tree.FunctionBodyBlockNode.apply(FunctionBodyBlockNode.java:59)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createFunctionNode(BLangNodeTransformer.java:1206)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:1163)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:418)
	at io.ballerinalang.compiler.syntax.tree.FunctionDefinitionNode.apply(FunctionDefinitionNode.java:71)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:505)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:418)
	at io.ballerinalang.compiler.syntax.tree.ModulePartNode.apply(ModulePartNode.java:54)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.accept(BLangNodeTransformer.java:443)
	at org.wso2.ballerinalang.compiler.parser.Parser.generateCompilationUnitNew(Parser.java:144)
	at org.wso2.ballerinalang.compiler.parser.Parser.parseNew(Parser.java:117)
	at org.wso2.ballerinalang.compiler.PackageLoader.parse(PackageLoader.java:500)
	at org.wso2.ballerinalang.compiler.PackageLoader.loadEntryPackage(PackageLoader.java:380)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackages(Compiler.java:156)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackages(Compiler.java:137)
	at org.ballerinalang.langserver.compiler.LSModuleCompiler.compilePackagesSafe(LSModuleCompiler.java:291)
	... 19 more
```

**Comment by m36dot** [comment]
the code is recovered as 
```
public function testFunction() {
    int[] numberList = [1, 2, 3];

    foreach i
}

function doSomeTask()  MISSING[in]{
    // todo: here goes a certain task
} MISSING[{] MISSING[}] MISSING[}]
```

Since functional-binding-patterns are not yet implemented in the node transformer there is a class cast exception.

**Comment by rdhananjaya** [comment]
Swanlake release will not support functional binding patterns/functional constructors. Pleae refer the spec issue https://github.com/ballerina-platform/ballerina-spec/issues/497

**Comment by KRVPerera** [comment]
Similar issue 

```Ballerina
public funcion main() {}
```

```LogTalk
[2020-09-11 10:01:29,740] SEVERE {b7a.log.crash} - io.ballerinalang.compiler.syntax.tree.FunctionalBindingPatternNode cannot be cast to io.ballerinalang.compiler.syntax.tree.CaptureBindingPatternNode 
java.lang.ClassCastException: io.ballerinalang.compiler.syntax.tree.FunctionalBindingPatternNode cannot be cast to io.ballerinalang.compiler.syntax.tree.CaptureBindingPatternNode
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:527)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:417)
        at io.ballerinalang.compiler.syntax.tree.ModuleVariableDeclarationNode.apply(ModuleVariableDeclarationNode.java:67)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:512)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:417)
        at io.ballerinalang.compiler.syntax.tree.ModulePartNode.apply(ModulePartNode.java:54)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.accept(BLangNodeTransformer.java:442)
        at org.wso2.ballerinalang.compiler.parser.Parser.generateCompilationUnitNew(Parser.java:144)
        at org.wso2.ballerinalang.compiler.parser.Parser.parseNew(Parser.java:117)
        at org.wso2.ballerinalang.compiler.PackageLoader.parse(PackageLoader.java:500)
        at org.wso2.ballerinalang.compiler.PackageLoader.loadEntryPackage(PackageLoader.java:380)
        at org.wso2.ballerinalang.compiler.Compiler.compilePackages(Compiler.java:156)
        at org.wso2.ballerinalang.compiler.Compiler.compilePackage(Compiler.java:174)
        at org.wso2.ballerinalang.compiler.Compiler.compile(Compiler.java:91)
        at org.wso2.ballerinalang.compiler.Compiler.build(Compiler.java:99)
        at org.ballerinalang.packerina.task.CompileTask.execute(CompileTask.java:57)
        at org.ballerinalang.packerina.TaskExecutor.executeTasks(TaskExecutor.java:38)
        at org.ballerinalang.packerina.cmd.RunCommand.execute(RunCommand.java:299)
        at java.util.Optional.ifPresent(Optional.java:159)
        at org.ballerinalang.tool.Main.main(Main.java:57)
```

**Comment by rdulmina** [comment]
FunctionalBindingPattern is removed from the master and the original issue is fixed. How ever there is still a bad sad error for 

`public funcion main() {}`

Here is the log from the master

```
[2020-09-11 10:20:31,973] SEVERE {b7a.log.crash} - io.ballerinalang.compiler.syntax.tree.ErrorBindingPatternNode cannot be cast to io.ballerinalang.compiler.syntax.tree.CaptureBindingPatternNode 
java.lang.ClassCastException: io.ballerinalang.compiler.syntax.tree.ErrorBindingPatternNode cannot be cast to io.ballerinalang.compiler.syntax.tree.CaptureBindingPatternNode
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:530)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:420)
	at io.ballerinalang.compiler.syntax.tree.ModuleVariableDeclarationNode.apply(ModuleVariableDeclarationNode.java:67)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:515)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:420)
	at io.ballerinalang.compiler.syntax.tree.ModulePartNode.apply(ModulePartNode.java:54)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.accept(BLangNodeTransformer.java:445)
	at org.wso2.ballerinalang.compiler.parser.Parser.generateCompilationUnitNew(Parser.java:144)
	at org.wso2.ballerinalang.compiler.parser.Parser.parseNew(Parser.java:117)
	at org.wso2.ballerinalang.compiler.PackageLoader.parse(PackageLoader.java:500)
	at org.wso2.ballerinalang.compiler.PackageLoader.loadEntryPackage(PackageLoader.java:380)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackages(Compiler.java:156)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackage(Compiler.java:174)
	at org.wso2.ballerinalang.compiler.Compiler.compile(Compiler.java:91)
	at org.wso2.ballerinalang.compiler.Compiler.build(Compiler.java:99)
	at org.ballerinalang.packerina.task.CompileTask.execute(CompileTask.java:57)
	at org.ballerinalang.packerina.TaskExecutor.executeTasks(TaskExecutor.java:38)
	at org.ballerinalang.packerina.cmd.RunCommand.execute(RunCommand.java:299)
	at java.util.Optional.ifPresent(Optional.java:159)
	at org.ballerinalang.tool.Main.main(Main.java:57)
```

**Comment by KRVPerera** [comment]
Please check below example also. 
```ErrorBindingPatternNode cannot be cast to CaptureBindingPatternNode```

```Ballerina
    var x = object {
        int n = 0;
        function inc() returns int {
            return self.n; // <-- needed for crash reprduce
        }; // `;` <-- needed for crash reproduce

        fution init() { // `fution`
        }
    };
```

**Comment by rdulmina** [comment]
Fixed in  #26949
